### DISCUSSION TITLE
Save UI as separate file

### DISCUSSION THREAD

**Discussion by LalFX** [israel-dryer/ttkbootstrap#489] [Q&A] [answered]
Is it possible to save the UI as a separate file? 

The issue is that if we have a fairly complex UI, the UI code is so long. It would be great if we could separate the UI from the Logic.

**Comment by ghost** [comment, marked as answer]
One way I have found to do this is to create class with all the form layout, and implement get and set methods to load the form and retrieve the data. Something like this:


```
from typing import Any

from ttkbootstrap import Entry, Label, StringVar, Toplevel  # type: ignore

class MY_UI(Toplevel):
    def __init__(self, *args: Any, **kwargs: Any) -> None:
        super().__init__(*args, themename="darkly", **kwargs)  # type: ignore

        self.label1: Label = Label(self, text="Entry one")
        self.label1.grid(row=0, column=0, padx=10, pady=10)

        self.entry1_value: StringVar = StringVar()
        self.entry1: Entry = Entry(self, textvariable=self.entry1_value)
        self.entry1.grid(row=0, column=1, padx=10, pady=10)

        self.label2: Label = Label(self, text="Entry two")
        self.label2.grid(row=1, column=0, padx=10, pady=10)

        self.entry2_value: StringVar = StringVar()
        self.entry2: Entry = Entry(self, textvariable=self.entry2_value)
        self.entry2.grid(row=1, column=1, padx=10, pady=10)

    def get(self) -> tuple[str, str]:
        return (self.entry1_value.get(), self.entry2_value.get())

    def set(self, value1: str, value2: str) -> None:
        print("set")
        self.entry1_value.set(value1)
        self.entry2_value.set(value2)
```

Then you can use it like any other object, using, for example:

my_ui : MY_UI = MY_UI()
my_ui.set(...) or my_ui,get() etc.

The mainloop would have to be in another window.

If the above code was saved in `my_ui.py` you would import it with:

`from my_ui import MY_UI`

**Reply by LalFX** [reply]
Thanks - this looks promising. Let me try it out and let you know.

**Comment by rdbende** [comment]
> Is it possible to save the UI as a separate file?
>
> The issue is that if we have a fairly complex UI, the UI code is so long. It would be great if we could separate the UI from the Logic.

This is not something that should be done by a style library like `ttkbootstrap`. For what you want there's [pygubu](https://github.com/alejandroautalan/pygubu) and [pygubu-designer](https://github.com/alejandroautalan/pygubu-designer)

**Reply by LalFX** [reply]
I do develop in pygubu and its great, but would like to try developing using ttkbootstrap. 

If you have a complex UI the UI code is so long that looking through the logic gets tedious. If you can separate the UI from the logic - it makes development so much easier...

**Reply by rdbende** [reply]
> If you have a complex UI the UI code is so long that looking through the logic gets tedious. If you can separate the UI from the logic - it makes development so much easier...

That was pygugu is doing - separating the UI definition into a separate XML file. 
It now has support for plugins (https://github.com/alejandroautalan/pygubu/tree/master/src/pygubu/plugins), but unfortunately a ttkbootstrap plugin doesn't exist yet.
